source/extensions/quic_listeners/quiche/platform/quic_mem_slice_storage_impl.cc

```diff
@@ -0,0 +1,38 @@
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "extensions/quic_listeners/quiche/platform/quic_mem_slice_storage_impl.h"
+
+#include "quiche/quic/core/quic_utils.h"
+
+namespace quic {
+
+QuicMemSliceStorageImpl::QuicMemSliceStorageImpl(const struct iovec* iov, int iov_count,
+                                                 QuicBufferAllocator* allocator,
+                                                 const QuicByteCount max_slice_len) {
+  if (iov == nullptr) {
+    return;
+  }
+  QuicByteCount write_len = 0;
+  for (int i = 0; i < iov_count; ++i) {
+    write_len += iov[i].iov_len;
+  }
+  size_t io_offset = 0;
+  while (io_offset < write_len) {
+    size_t slice_len = std::min(write_len - io_offset, max_slice_len);
+    char* dst = allocator->New(slice_len + sizeof(Envoy::Buffer::BufferFragmentImpl));
```

Comment:
See my comment on QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length). It looks like we should put  bundling+allocation+constructing fragment into 1 place and use it from both places.